IA-4285: Adding a FHIR API to retrieve org units

madewulf opened the pull request:
Add a FHIR api to retrieve org units
 
[IA-4285]  create-a-fhir-retrieval-api-for-org-units

## Self proofreading checklist

- [x] Did I use eslint and ruff formatters?
- [x] Is my code clear enough and well documented?
- [x] Are there enough tests?
- [x] Documentation has been included (for new feature)

## Doc

In the new readme in /api/fhir

## Changes

See the readme

## How to test

See the readme

## Notes

We should test now the integration with an external system, that was not done yet. 


## Follow the Conventional Commits specification

The **merge message** of a pull request must follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) specification.

This convention helps to automatically generate release notes.

Use lowercase for consistency.

[Example](https://github.com/BLSQ/iaso/commit/8b8d7d3064138c1e57878f17b4eb922516ab0112):

```
fix: empty instance pop up

Refs: IA-3665
```

Note that the Jira reference is preceded by a _line break_.

Both the line break and the Jira reference are entered in the _Add an optional extended description…_ field.


[IA-4285]: https://bluesquare.atlassian.net/browse/IA-4285?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

madewulf committed:
Adding a FHIR API to retrieve org units

madewulf committed:
Changing the mapping of IASO status to FHIR status to be more logical

madewulf committed:
Fixing tests

madewulf committed:
Fix tests

madewulf committed:
Switching to openiaso.com for the domain used in identifiers

kemar commented on iaso/api/fhir/location.py:
```diff
@@ -0,0 +1,304 @@
+            page_size = int(request.query_params.get("_count", 20))
+            page_size = min(page_size, 100)  # Max 100
+        except (TypeError, ValueError):
+            page_size = 20
+
+        try:
+            offset = int(request.query_params.get("_skip", 0))
+        except (TypeError, ValueError):
+            offset = 0
+
+        return page_size, offset
+
+    def _build_pagination_links(self, request, total_count: int) -> List[Dict[str, str]]:
+        """Build FHIR pagination links"""
+        links = []
+
+        # Get pagination parameters
+        page_size, offset = self._get_pagination_params(request)
+
+        # Self link
+        self_url = request.build_absolute_uri().split("?")[0]
+        if request.query_params:
+            self_url += f"?{request.query_params.urlencode()}"
+        links.append({"relation": "self", "url": self_url})
+
+        # Next link
+        if offset + page_size < total_count:
+            next_params = request.query_params.copy()
+            next_params["_skip"] = offset + page_size
+            next_url = f"{request.build_absolute_uri().split('?')[0]}?{next_params.urlencode()}"
+            links.append({"relation": "next", "url": next_url})
+
+        # Previous link
+        if offset > 0:
+            prev_params = request.query_params.copy()
+            prev_params["_skip"] = max(0, offset - page_size)
+            prev_url = f"{request.build_absolute_uri().split('?')[0]}?{prev_params.urlencode()}"
+            links.append({"relation": "previous", "url": prev_url})
+
+        return links
```
This whole pagination logic could have been implemented as a `LimitOffsetPagination` class:

```python
class FHIRPaginator(pagination.LimitOffsetPagination):
    default_limit = 20
    limit_query_param = "_count"
    limit_query_description = 'Number of results to return per page.'
    offset_query_param = "_skip"
    offset_query_description = "The initial index from which to return the results."
    max_limit = 100

    …

```

Avoiding mixing "views" and "pagination" logic in the same place could provide some clarity.

References:

- https://www.django-rest-framework.org/api-guide/pagination/#limitoffsetpagination
- https://github.com/encode/django-rest-framework/blob/e454758fb6edf1dcf5aa5417a388b940c871469c/rest_framework/pagination.py#L373

kemar commented on iaso/api/fhir/location.py:
```diff
@@ -0,0 +1,304 @@
+            return Response(error_response, status=status.HTTP_404_NOT_FOUND)
+
+    @action(detail=False, methods=["get"])
+    def metadata(self, request):
```
Is getting `metadata` via a `GET` method a FHIR API requirement?

I'm asking because there is an existing [metadata mechanism in DRF](https://www.django-rest-framework.org/api-guide/metadata/) but it relies on the `OPTIONS` HTTP method.

kemar commented on iaso/api/fhir/location.py:
I would have put the filter and the permissions in their own files (`filters.py` and `permissions.py`) for consistency as well as renaming this one to `views.py`

kemar commented on iaso/api/fhir/tests.py:
Why not put this file with the others into `/iaso/tests/`?

Is this on purpose?

kemar commented on iaso/urls.py:
```diff
@@ -261,6 +261,7 @@ def append_datasources_subresource(viewset, resource_name, urlpatterns):
     path("storages/<str:storage_type>/<str:storage_customer_chosen_id>/logs", logs_per_device),
     path("workflows/export/<workflow_id>/", export_workflow, name="export_workflow"),
     path("workflows/import/", import_workflow, name="import_workflow"),
+    path("fhir/", include("iaso.api.fhir.urls")),
```
The way you implemented that makes the route unavailable in the default list. Is this done on purpose?

![a](https://github.com/user-attachments/assets/96ddfb98-14f2-421c-87c8-5c8b8bc8e6eb)

kemar reviewed (changes requested):
Looks almost good to me.

I have a few suggestions that could put the iaso codebase towards more consistency though.

madewulf committed:
refactor: split FHIR module into filters/permissions/pagination/views; move tests

- Extract FHIRLocationFilter to filters.py
- Extract FHIRLocationPermission to permissions.py
- Create FHIRPaginator (LimitOffsetPagination) in pagination.py using _count/_skip
- Rename location.py to views.py; viewset now uses pagination_class
- Register FHIRLocationViewSet in main router (visible in DRF browsable API)
- Move tests to iaso/tests/api/fhir/test_location.py
- Update README: expand FHIR acronym and link to spec

quang-le committed:
Merge branch 'develop' into IA-4285-create-a-fhir-retrieval-api-for-org-units

quang-le committed:
IA-4285 fix permission + failing test

quang-le committed:
IA-4285: add tests, update perms

quang-le committed:
IA_4285: add openAPI annotations

quang-le committed:
linter

quang-le committed:
IA-4285: add FHIR section to swagger

hugo093 commented on iaso/urls.py:
```diff
@@ -293,7 +294,7 @@
 )
 router.register(r"mobile/validation-workflows", ValidationWorkflowMobileViewSet, basename="mobile_validation_workflows")
-
+router.register(r"fhir/Location", FHIRLocationViewSet, basename="fhir-location")
```
```suggestion
router.register(r"fhir/location", FHIRLocationViewSet, basename="fhir-location")
```

hugo093 commented on iaso/api/fhir/filters.py:
```diff
@@ -0,0 +1,30 @@
+from django.db import models as django_models
```
```suggestion
from django.db.models import Q
```

hugo093 commented on iaso/api/fhir/filters.py:
```diff
@@ -0,0 +1,30 @@
+    def filter_identifier(self, queryset, name, value):
+        return queryset.filter(
+            django_models.Q(source_ref=value) | django_models.Q(uuid=value) | django_models.Q(aliases__contains=[value])
+        )
```
```suggestion
        if value:
            return queryset.filter(
                django_models.Q(source_ref=value) | django_models.Q(uuid=value) | django_models.Q(aliases__contains=[value])
            )
        return queryset
```

hugo093 commented on iaso/api/fhir/README.md:
```diff
@@ -0,0 +1,259 @@
+
+| Method | Endpoint | Description |
+|--------|----------|-------------|
+| GET | `/fhir/Location/` | List all locations (returns FHIR Bundle) |
+| GET | `/fhir/Location/{id}/` | Get a specific location by ID |
+| GET | `/fhir/Location/{id}/children/` | Get child locations of a specific location |
+| GET | `/fhir/Location/metadata/` | Get FHIR CapabilityStatement |
```
```suggestion
| GET | `/fhir/location/` | List all locations (returns FHIR Bundle) |
| GET | `/fhir/location/{id}/` | Get a specific location by ID |
| GET | `/fhir/location/{id}/children/` | Get child locations of a specific location |
| GET | `/fhir/location/metadata/` | Get FHIR CapabilityStatement |
```

hugo093 commented on iaso/api/fhir/README.md:
````diff
@@ -0,0 +1,259 @@
+### List All Locations
+
+```http
+GET /fhir/Location/
````
```suggestion
GET /fhir/location/
```

hugo093 commented on iaso/api/fhir/README.md:
````diff
@@ -0,0 +1,259 @@
+### Get Specific Location
+
+```http
+GET /fhir/Location/123/
````
```suggestion
GET /fhir/location/123/
```

hugo093 commented on iaso/api/fhir/README.md:
````diff
@@ -0,0 +1,259 @@
+### Search by Name
+
+```http
+GET /fhir/Location/?name=hospital
````
```suggestion
GET /fhir/location/?name=hospital
```

hugo093 commented on iaso/api/fhir/README.md:
````diff
@@ -0,0 +1,259 @@
+### Search by Status
+
+```http
+GET /fhir/Location/?status=active
````
```suggestion
GET /fhir/location/?status=active
```

hugo093 commented on iaso/api/fhir/README.md:
````diff
@@ -0,0 +1,259 @@
+### Get Child Locations
+
+```http
+GET /fhir/Location/122/children/
````
```suggestion
GET /fhir/location/122/children/
```

hugo093 commented on iaso/api/fhir/README.md:
````diff
@@ -0,0 +1,259 @@
+### Pagination
+
+```http
+GET /fhir/Location/?_count=25&_skip=50
````
```suggestion
GET /fhir/location/?_count=25&_skip=50
```

hugo093 commented on iaso/api/fhir/urls.py:
```diff
@@ -0,0 +1,17 @@
+"""
```
it should either be deleted , or iaso.urls should be adapted to include that , and not the viewsets directly

hugo093 commented on iaso/api/fhir/pagination.py:
```diff
@@ -0,0 +1,45 @@
+        return {
+            "resourceType": "Bundle",
+            "id": bundle_id,
+            "meta": {"lastUpdated": "2024-01-01T00:00:00Z"},
```
oupsie?

hugo093 commented on iaso/api/fhir/pagination.py:
```diff
@@ -0,0 +1,45 @@
+    max_limit = 100
+
+    def get_paginated_response(self, data):
+        base_url = self.request.build_absolute_uri().split("?")[0].rstrip("/")
```
1. should use urllib to do that
2. It's already handled in get_fhir_bundle, remove it

hugo093 committed:
feat: some fixes, added https, remove hardcoded dates, improve a bit swagger compliance

quang-le committed:
IA-4285: add doc comment
